acceptancetests/jujupy/stream_server.py

```diff
@@ -0,0 +1,293 @@
+        ip_address, port = _get_server_address(server)
+        address = 'http://{}:{}'.format(ip_address, port)
+        server_process = multiprocessing.Process(
+            target=_http_worker,
+            args=(server, self.base_dir),
+            name='SimlestreamServer')
+        try:
+            log.info('Starting stream server at: {}'.format(address))
+            multiprocessing.log_to_stderr(logging.DEBUG)
+            server_process.start()
+            yield address
+        finally:
+            log.info('Terminating stream server')
+            server_process.terminate()
+            server_process.join()
+
+
+def agent_tgz_from_juju_binary(juju_bin_path, tmp_dir, series=None):
+    """
+    Create agent tarball with jujud found with provided juju binary.
+
+    Search the location where `juju_bin_path` resides to attempt to find a
+    jujud in the same location.
+
+    :param juju_bin_path: The path to the juju bin in use.
+    :param tmp_dir: Location to store the generated agent file.
+    :param series: String series to use instead of that of the passed binary.
+      Allows one to overwrite the series of the juju client.
+    :returns: String path to generated
+    """
+    _series_lookup = dict(
+        artful='ubuntu',
+        bionic='ubuntu',
+        trusty='ubuntu',
+        xenial='ubuntu',
+        win2012='win2012',
+        win2012hv='win2012',
+        win2012hvr2='win2012',
+        win2012r2='win2012',
+        win2016='win2012',
+        win2016nano='win2012',
+        win7='win2012',
+        win81='win2012',
+        win8='win2012',
+        win10='win2012',
+    )
+
+    bin_dir = os.path.dirname(juju_bin_path)
+    try:
+        jujud_path = os.path.join(
+            bin_dir,
+            [f for f in os.listdir(bin_dir) if f == 'jujud'][0])
+    except IndexError:
+        raise RuntimeError('Unable to find jujud binary in {}'.format(bin_dir))
+
+    try:
+        version_output = subprocess.check_output(
+            [jujud_path, 'version']).rstrip('\n')
+        version, bin_series, arch = version_output.split('-')
+        bin_agent_series = _series_lookup.get(series, series)
+    except subprocess.CalledProcessError as e:
+        raise RuntimeError(
+            'Unable to query jujud for version details: {}'.format(e))
+    except IndexError:
+        raise RuntimeError(
+            'Unable to determine version, series and arch from version '
+            'string: {}'.format(version_output))
+
+    agent_tgz_name = 'juju-{version}-{series}-{arch}.tgz'.format(
+        version=version,
+        series=series if series else bin_agent_series,
+        arch=arch
+    )
+
+    log.debug('Creating agent file: {}'.format(agent_tgz_name))
+    tgz_path = os.path.join(tmp_dir, agent_tgz_name)
+    with tarfile.open(tgz_path, 'w:gz') as tar:
+        tar.add(jujud_path, arcname='jujud')
+
+    return tgz_path
+
+
+def _generate_product_json(content_id, version, arch, series, agent_tgz_path):
```

Comment:
There may also be simplestreams libraries as part of the simplestreams project we can import from instead of using bespoke code here